engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala

```diff
@@ -132,28 +132,26 @@ class DevelopmentDeploymentManager(actorSystem: ActorSystem)
 
   override def customActions: List[CustomAction] = customActionStatusMapping.keys.toList
 
-  override def invokeCustomAction(actionRequest: CustomActionRequest, canonicalProcess: CanonicalProcess): Future[Either[CustomActionError, CustomActionResult]] =
-    Future.successful(
-      customActions
-        .find(_.name.equals(actionRequest.name))
-        .map(customAction => {
-          val processName = actionRequest.processVersion.processName
-          val statusDetails = memory.getOrElse(processName, createAndSaveProcessState(NotDeployed, actionRequest.processVersion))
-
-          if (customAction.allowedStateStatusNames.contains(statusDetails.status.name)) {
-            customActionStatusMapping
-              .get(customAction)
-              .map { status =>
-                asyncChangeState(processName, status)
-                Right(CustomActionResult(actionRequest, s"Done ${actionRequest.name}"))
-              }
-              .getOrElse(Left(CustomActionInvalidStatus(actionRequest, statusDetails.status.name)))
-          } else {
-            Left(CustomActionInvalidStatus(actionRequest, statusDetails.status.name))
-          }
-        })
-        .getOrElse(Left(CustomActionNotImplemented(actionRequest)))
-    )
+  override def invokeCustomAction(actionRequest: CustomActionRequest, canonicalProcess: CanonicalProcess): Future[CustomActionResult] =
+    customActions
+      .find(_.name.equals(actionRequest.name))
+      .map(customAction => {
+        val processName = actionRequest.processVersion.processName
+        val statusDetails = memory.getOrElse(processName, createAndSaveProcessState(NotDeployed, actionRequest.processVersion))
+
+        if (customAction.allowedStateStatusNames.contains(statusDetails.status.name)) {
+          customActionStatusMapping
+            .get(customAction)
+            .map { status =>
+              asyncChangeState(processName, status)
+              Future.successful(CustomActionResult(actionRequest, s"Done ${actionRequest.name}"))
+            }
+            .getOrElse(Future.failed(CustomActionInvalidStatus(actionRequest, statusDetails.status.name)))
```

Comment:
This is the code that I mentioned to you that we should change. Take a look - in our design, `CustomActionInvokerServiceImpl` checks for which status, action can be invoked. Thanks to that we shouldn't check the same in our custom action implementation - it is already done. This means classes like `CustomActionInvalidStatus` shouldn't be a part of the API. It should be an internal exception. IMO we should change this developer example so `CustomActionInvalidStatus` won't be the part of API anymore

Reply:
Fixed. All general customAction checks are performed in `CustomActionInvokerServiceImpl` and `DevelopmentDeploymentManagerProvider` is responsible only for what happens within the action.